Mathlib/Data/List/Basic.lean

```diff
@@ -1010,6 +1010,43 @@ theorem zipWith_flip (f : α → β → γ) : ∀ as bs, zipWith (flip f) bs as
 
 /-! ### take, drop -/
 
+@[simp] lemma take_eq_self (x : List α) n : x.take n = x ↔ x.length ≤ n :=
+  ⟨fun h ↦ by rw [← h]; simp; omega, take_of_length_le⟩
+
+@[simp] lemma take_self_eq (x : List α) n : x = x.take n ↔ x.length ≤ n := by
+  rw [Eq.comm, take_eq_self]
+
+@[simp] lemma take_eq_left {x y : List α} n :
+  (x ++ y).take n = x.take n ↔ y = [] ∨ n ≤ x.length := by
+  constructor
```

Comment:
I have the glaring feeling this proof can be optimised. I'm not familiar enough with the library around lists to quickly suggest how, though. Perhaps another reviewer can provide some feedback.

Reply:
Your feeling was right